lib/spack/spack/solver/asp.py

```diff
@@ -824,28 +834,56 @@ def _lock_prefix_from_cache_path(self, cache_path: str):
             spack.util.hash.b32_hash(cache_path), spack.util.crypto.bit_length(sys.maxsize)
         )
 
+    def _zip_cache_entry(self, cache_data: str):
+        """Compress cache entries via gzip"""
+        return gzip.compress(cache_data.encode("utf-8"))
+
+    def _unzip_cache_entry(self, cache_data: bytes):
+        """Decompresses cache entries, which are gzipped by default"""
+        return gzip.decompress(cache_data).decode("utf-8")
+
+    def _synchronize_manifest(self, old, new):
+        """Ensures that cache manifest is synchronized between the original
+        and temporary file created by the file cache write transaction manager
+
+        Returns both stream cursors to beginning of each stream
+        """
+        # synchronize manifests as old will be deleted by
+        # filecache once the context manager exits
+        # skip the manifest header data
+        old.readline()
+        cache_context = old.read()
+        old.seek(0, io.SEEK_SET)
+        # write an empty to line for manifest header
+        new.write("\n")
+        new.write(cache_context)
+        new.seek(0, io.SEEK_SET)
+
     def flush_manifest(self):
         """Updates the concretization cache manifest file after a cache write operation
         Updates the current byte count and entry counts and writes to the head of the
         manifest file"""
         manifest_file = self.root / self._cache_manifest
         manifest_file.touch(exist_ok=True)
-        with open(manifest_file, "r+", encoding="utf-8") as f:
+        with self._fc.write_transaction(self._cache_manifest) as (old, new):
+            self._synchronize_manifest(old, new)
             # check if manifest is empty
-            count, cache_bytes = self._extract_cache_metadata(f)
+            # now all operations can be perofrmed
+            count, cache_bytes = self._extract_cache_metadata(old)
             if not count or not cache_bytes:
                 # cache is unintialized
                 count = 0
                 cache_bytes = 0
-            f.seek(0, io.SEEK_END)
+            new.seek(0, io.SEEK_END)
             for manifest_update in self._manifest_queue:
                 entry_path, entry_bytes = manifest_update
                 count += 1
                 cache_bytes += entry_bytes
-                f.write(f"{entry_path.name} {entry_bytes}")
-            f.seek(0, io.SEEK_SET)
-            new_stats = f"{int(count)+1} {int(cache_bytes)}\n"
-            f.write(new_stats)
+                new.write(f"{entry_path.name} {entry_bytes}\n")
+            new.seek(0, io.SEEK_SET)
+            new_stats = f"{count} {cache_bytes}\n"
```

Comment:
This in-place mutation does not work because line length is variable. If it increases, you overwrite the newline character, and the file gets corrupted. I think you should either work with a binary file (`struct` module), or rewrite the entire file (probably what you wanted to avoid).

Note that if the idea is to do a constant time write operation (update header & append single entry), then `FileCache` is the wrong choice, since it copies the file to a temporary which you then mutate in-place. If you want to do things in-place, you could work with the lock file directly.

Reply:
Also what's missing is some version field in the header to describe the manifest/cache file format, which is always helpful in case we make a forward incompatible change.

Reply:
> what's missing is some version field in the header to describe the metadata file format

Good idea, adding now. 

> This in-place mutation does not work because line length is variable.

🤦 

> Note that if the idea is to do a constant time write operation (update header & append single entry), then FileCache is the wrong choice, since it copies the file to a temporary which you then mutate in-place.

Good point, I didn't consider this when I refactored to using the FileCache, I may as well just avoid the in place write. That said, this outcome of this response conditions on our overall decision on whether or not to actually use the `FileCache`. 

Either way, I agree, it should all be binary.